pkg/xds/server/components.go

```diff
@@ -271,6 +271,20 @@ func DefaultDataplaneSyncTracker(rt core_runtime.Runtime, reconciler, ingressRec
 					return err
 				}
 
+				// Need to get dataplane again after counting hash. Because the following situation is possible:
+				// 1. Get Dataplane(version 1)
+				// 2. Dataplane is externally updated to version 2
+				// 3. Calculate hash in meshSnapshotCache.GetHash. This function list all resources and hash will be built based on Dataplane(version 2)
+				// 4. Use Dataplane(version 1) when generate Envoy resources
+				//
+				// So update from version 1 to version 2 is effectively lost.
+				if err := rt.ReadOnlyResourceManager().Get(ctx, dataplane, core_store.GetBy(key)); err != nil {
+					if core_store.IsResourceNotFound(err) {
+						return reconciler.Clear(&proxyID)
+					}
+					return err
+				}
```

Comment:
to "port" my comment. I think it won't solve the problem you either have to use dataplane that was used to build hash or use `ResourceManager()` (non cached)

Reply:
We are not doing `ResolveAddress` after we fetch it.

Do we even need to resolve it twice? Can we skip first fetch?
If we do:
1) check hash mesh
2) fetch the dataplane

instead of
1) fetch the dataplane
2) check hash mesh

We are reducing A LOT of queries to the database instead of introducing A LOT more queries because of 2 fetches.

Also I'm wondering now. What is happening when we are disconnecting the Dataplane and we don't have `reconciler.Clear(&proxyID)` line. This is not guaranteed that `reconciler.Clear(&proxyID)` will be executed. Is the snapshot for given dataplane stuck indefinitely in the cache (essentially a memory leak) or is it cleaned up?